Mathlib/Data/Finset/Antidiagonal.lean

```diff
@@ -0,0 +1,185 @@
+/-
+Copyright (c) 2023 Antoine Chambert-Loir and María Inés de Frutos-Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos-Fernández, Bhavik Mehta
+-/
+
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Finsupp.Defs
+import Mathlib.Data.Finsupp.Interval
+import Mathlib.Algebra.Order.Sub.Defs
+
+/-! # Antidiagonal with values in general types
+
+Let `n : μ`, where `μ` is a canonically ordered add monoid with locally finite order.
+(For example, one may take `μ` to be `ℕ`, more generally `σ →₀ ℕ`.)
+
+* For `s : Finset ι`, we define `Finset.piAntidiagonal s n` as the `Finset (ι → μ)`
+of functions with support in `s` whose sum is equal to `n`.
+
+* We define `Finset.antidiagonal n : Finset (μ × μ)` of pairs adding to `n`.
+
+These definitions generalize `Finset.Nat.antidiagonal` and `Finsupp.antidiagonal`
+which are defined in `Mathlib.Data.Finset.NatAntidiagonal`
+and `Mathlib.Data.Finsupp.Antidiagonal` and we make the comparisons there.
+
+This definition does not exactly match with that of `Multiset.antidiagonal` which is
+defined in `Mathlib.Data.Multiset.Antidiagonal`, because of the multiplicities.
+Indeed, by counting multiplicities, `Multiset α` is equivalent to `α →₀ ℕ`,
+but `Finset.antidiagonal` and `Multiset.antidiagonal` will return different objects.
+For example, for `s : Multiset ℕ := {0,0,0}`, `Multiset.antidiagonal s` has 8 elements
+but `Finset.antidiagonal s` has only 4.
+
+```lean
+def s : Multiset ℕ := {0, 0, 0}
+#eval (Finset.antidiagonal s).card -- 4
+#eval Multiset.card (Multiset.antidiagonal s) -- 8
+```
+
+This is mostly taken from a file of Bhavik Mehta,
+Archive/Wiedijk100Theorems/Partition.lean,
+where `piAntidiagonal` was called `cut`
```

Comment:
I think you should remove this comment entirely, and replace `cut` with your new definition!

Reply:
You mean, in the Archive file ?

Reply:
Yes!

Reply:
I did that, but there was a conflict of the decidable instances. I managed to pass through, but this is ugly and  I don't believe we wish to leave that as is.
(In presence of `hf : f in piAntidiagonal s n` with bad decidable instances, I added `suffices : hf' : f in piAntidiagonal s n`, which has the good instances and which  I proved using `convert hf`.

Reply:
I could adjust the Archive file using piAntidiagonal.